pkg/tuf/client.go

```diff
@@ -15,137 +15,184 @@
+// loadMetadata controls if the client actually should perform a TUF refresh.
+// The TUF specification mandates so, but for certain Sigstore clients, it
+// may be beneficial to rely on the cache, or in air-gapped deployments it
+// it may not even be possible.
+func (c *Client) loadMetadata() error {
+	// Load the metadata into memory and verify it
+	if err := c.up.Refresh(); err != nil {
+		// this is most likely due to the lack of metadata files
+		// on disk. Perform a full update and return.
+		return c.Refresh()
 	}
 
-	trustedroot, ok := tufMetaMap[TrustedRootTUFPath]
-	if ok {
-		if ok, _ := validTarget(trustedrootMeta, trustedroot); ok {
-			return trustedroot, nil
+	var tm = c.up.GetTrustedMetadataSet()
+	if c.opts.ForceCache {
+		// Use cache until it expires
+		if tm.Timestamp.Signed.IsExpired(time.Now()) {
+			return c.Refresh()
 		}
-	}
 
-	// What's on disk didn't match, so download from TUF remote (and cache it to disk)
-	writer := &Writer{
-		Bytes: make([]byte, 0),
-	}
+		// Cache not expired, return
+		return nil
+	} else if c.opts.CacheValidity > 0 {
+		// Use cached metadata for up to CacheValidity days.
+		// This is a bit of an hack, as we don't know when the
+		// last the it was updated, fallback to check the
+		// modification time of timestamp.json
+		if tm.Timestamp.Signed.IsExpired(time.Now()) {
+			// Always update if the timestamp is expired
+			return c.Refresh()
+		}
 
-	err = tufClient.Download(TrustedRootTUFPath, writer)
-	if err != nil {
-		return nil, err
-	}
+		var p = filepath.Join(
+			c.opts.CachePath,
+			URLToPath(c.opts.RepositoryBaseURL),
+			"timestamp.json",
+		)
+		fi, err := os.Stat(p)
+		if err != nil {
+			// Failed to get info on the file, fall back
+			// and update if needed
+			return c.Refresh()
+		}
 
-	err = fileJSONStore.SetMeta(TrustedRootTUFPath, writer.Bytes)
-	if err != nil {
-		return nil, err
+		if fi.ModTime().After(time.Now().Add(
+			time.Duration(-24*c.opts.CacheValidity) * time.Hour)) {
+			// No need to update
+			return nil
+		}
+		// A TUF client refresh will now happen (c.Refresh),
+		// update the mod time for the timestamp.
+		//
+		// Ignore the error here, there is no need to fail
+		// operation only because the file's metadata could
+		// not be updated
+		//nolint:errcheck
+		os.Chtimes(p, time.Now(), time.Now())
 	}
 
-	return writer.Bytes, nil
+	return c.Refresh()
 }
 
-func checkEmbedded(tufRootURL string, fileJSONStore *filejsonstore.FileJSONStore) (json.RawMessage, error) {
-	embeddedRootPath := path.Join("repository", tufRootURL, RootTUFPath)
+// Refresh forces a refresh of the underlying TUF client.
+// As the tuf client does not support multiple refreshes during its
+// life-time, this will replace the TUF client with a new one.
+func (c *Client) Refresh() error {
+	var err error
 
-	root, err := embeddedRepos.ReadFile(embeddedRootPath)
+	c.up, err = updater.New(c.cfg)
 	if err != nil {
-		return nil, err
+		return err
 	}
+	return c.up.Refresh()
```

Comment:
How is concurrency handled with calling `Refresh` in paralell? By the underlying TUF repo?

Reply:
No, this should be documented as a method unsafe for concurrent use. For concurrency it appears the `LiveTrustedRoot` implements required serialization via a `Mutex`.

Reply:
I'll add an issue for that in go-tuf-metadata so we can work on it 👍